pkg/logs/sender/sender.go

```diff
@@ -6,220 +6,125 @@
-	Flush(ctx context.Context)
+type destinationState struct {
+	input       chan *message.Payload
+	destination client.Destination
+	stopChan    chan struct{}
 }
 
-// Sender contains all the logic to manage a stream of messages to destinations
-type Sender interface {
-	Start()
-	Stop()
-	Flush(ctx context.Context)
-}
-
-// SingleSender sends logs to different destinations.
-type SingleSender struct {
-	inputChan    chan *message.Message
-	outputChan   chan *message.Message
-	hasError     chan bool
+// Sender sends logs to different destinations.
+type Sender struct {
+	inputChan    chan *message.Payload
+	outputChan   chan *message.Payload
 	destinations *client.Destinations
-	strategy     Strategy
 	done         chan struct{}
-	lastError    error
-	trackErrors  bool
+	bufferSize   int
 }
 
-// NewSingleSender returns a new sender.
-func NewSingleSender(inputChan chan *message.Message, outputChan chan *message.Message, destinations *client.Destinations, strategy Strategy) *SingleSender {
-	return &SingleSender{
+// NewSender returns a new sender.
+func NewSender(inputChan chan *message.Payload, outputChan chan *message.Payload, destinations *client.Destinations, bufferSize int) *Sender {
+	return &Sender{
 		inputChan:    inputChan,
 		outputChan:   outputChan,
-		hasError:     make(chan bool, 1),
 		destinations: destinations,
-		strategy:     strategy,
 		done:         make(chan struct{}),
-		trackErrors:  false,
+		bufferSize:   bufferSize,
 	}
 }
 
 // Start starts the sender.
-func (s *SingleSender) Start() {
+func (s *Sender) Start() {
 	go s.run()
 }
 
 // Stop stops the sender,
 // this call blocks until inputChan is flushed
-func (s *SingleSender) Stop() {
+func (s *Sender) Stop() {
 	close(s.inputChan)
 	<-s.done
 }
 
-// Flush sends synchronously the messages that this sender has to send.
-func (s *SingleSender) Flush(ctx context.Context) {
-	s.strategy.Flush(ctx)
-}
+func (s *Sender) run() {
+	reliableDestinations := buildDestinationStates(s.destinations.Reliable, s.outputChan, s.bufferSize)
 
-func (s *SingleSender) run() {
-	defer func() {
-		s.done <- struct{}{}
-	}()
-	s.strategy.Send(s.inputChan, s.outputChan, s.send)
-}
+	sink := additionalDestinationsSink(s.bufferSize)
+	additionalDestinations := buildDestinationStates(s.destinations.Additionals, sink, s.bufferSize)
+
+	for payload := range s.inputChan {
 
-// send sends a payload to multiple destinations,
-// it will forever retry for the main destination unless the error is not retryable
-// and only try once for additional destinations.
-func (s *SingleSender) send(payload []byte) error {
-	for {
-		err := s.destinations.Main.Send(payload)
-		if err != nil {
-			if s.trackErrors && s.lastError == nil {
-				s.hasError <- true
+		sent := false
+		for !sent {
+			for _, destState := range reliableDestinations {
+				if !destState.destination.GetIsRetrying() {
+					destState.input <- payload
```

Comment:
I think this is making an implicit assumption that if `!GetIsRetrying()` then `input` is not full and this send will not block.  But those values are measured at different times, so that's not necessarily true.  I think that could cause a deadlock here, where this loop gets stuck because of one destination and does not service the others.

Reply:
This is intended to block and it handles 2 cases: 
1. All of the reliable destinations are retrying
 - in this case we want to poll until at least one is unstuck and block the whole pipeline
2. One destination is slower than the other
 - in this case it's input *could* fill up, and we want that to block and slow down the other destinations so they stay in sync.

Reply:
> This is intended to block and it handles 2 cases:
> ...
> * in this case we want to poll until at least one is unstuck and block the whole pipeline

But what is happening is that you're blocking until the first destination is unstuck (continuing to block if other destinations are unstuck).

You're also blocking if the first destination has a full channel but `!GetIsRetrying()`, and again not checking any of the other destinations.

I'm suggesting that in these cases it is *blocking* instead of *polling*.

> * in this case it's input _could_ fill up, and we want that to block and slow down the other destinations so they stay in sync.

This is a bit of a tricky one, but here too I think you want to poll instead of blocking the sending goroutine?  In general, any time this goroutine blocks sending to a channel, it runs the risk of blocking there forever (e.g., if that destination goes down) and never even looking at the other destinations, regardless of their status.  So it seems like it would _always_ want to use non-blocking sends.

Reply:
I just pushed an update that should address this. Ive added [DestinationSender](https://github.com/DataDog/datadog-agent/pull/10162/files#diff-4251c2e67dbd27f270a6e52213db5971149f7d0d7b9d9774263c3df3da8cd199R1) that exposes a blocking and non-blocking send method for each destination. 

The blocking send method is synchronized with the retry state and the blocking write to the input channel is cancelable if the retry state changes while it's blocking. Ive also added a [unit test](https://github.com/DataDog/datadog-agent/pull/10162/files#diff-3841552a484cca3c07cc5ac3aa401857ef2d6d0926d743979c078bf26449e4b3R42) for this scenario. Let me know what you think. This is indeed tricky.